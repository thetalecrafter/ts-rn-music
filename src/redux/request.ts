import { Dispatch, MiddlewareAPI } from 'redux'

const ISO_DATE = /^\d{4,}-\d{2}-\d{2}T\d{2}(:\d{2}:\d{2}(\.\d+)?([-+]\d{2}:\d{2})?Z?)?$/
export const reviver: any = (key: any, value: any) => {
  if (typeof value === 'string' && ISO_DATE.test(value)) {
    return new Date(value)
  }
  return value
}

export enum Status {
  Unsent = 'UNSENT',
  Loading = 'LOADING',
  Done = 'DONE',
  Error = 'ERROR'
}

export interface Request<Response> {
  error?: Error
  options?: RequestInit
  status: Status
  response?: Response
  url: string
}

export const request = <Response>(url: string, options?: RequestInit): Request<Response> => ({
  options,
  status: Status.Unsent,
  url
})

export default <S>(store: MiddlewareAPI<S>) => (next: Dispatch<S>) => (action: any) => {
  if (!action.request || action.request.status !== Status.Unsent) {
    return next(action)
  }

  next({
    ...action,
    request: {
      ...action.request,
      status: Status.Loading
    }
  })

  return fetch(action.request.url, action.request.options)
    .then((response) => response.text())
    .then((response) => ({
      ...action,
      request: {
        ...action.request,
        response: JSON.parse(response, reviver),
        status: Status.Done
      }
    }))
    .catch((error) => ({
      ...action,
      request: {
        ...action.request,
        error,
        status: Status.Error
      }
    }))
    .then((done) => store.dispatch(done))
}
