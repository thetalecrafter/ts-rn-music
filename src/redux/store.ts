import {
  applyMiddleware,
  bindActionCreators,
  combineReducers,
  createStore
} from 'redux'
import albumsActions from '../albums/actions'
import albumsReducer, { State as AlbumsState } from '../albums/reducer'
import songsActions from '../songs/actions'
import songsReducer, { State as SongsState } from '../songs/reducer'
import requestMiddleware from './request'

export interface State {
  albums: AlbumsState
  songs: SongsState
}

export const store = createStore<State>(combineReducers({
  albums: albumsReducer,
  songs: songsReducer
}), applyMiddleware(requestMiddleware))

export const actions = {
  albums: bindActionCreators(albumsActions, store.dispatch),
  songs: bindActionCreators(songsActions, store.dispatch)
}

export type Actions = typeof actions
