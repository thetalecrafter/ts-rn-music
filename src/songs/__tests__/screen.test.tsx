import { shallow, ShallowWrapper } from 'enzyme'
import * as React from 'react'
import { NavigationScreenProp } from 'react-navigation'
import { actions, store } from '../../redux/store'
import Screen from '../screen'
import Song from '../song'

describe('Songs screen', () => {
  const navigation = {
    dispatch: jest.fn(),
    goBack: jest.fn(),
    navigate: jest.fn(),
    setParams: jest.fn(),
    state: { params: {
      collectionId: 1024117649,
      collectionName: 'Over the Hedge (Music from the Motion Picture)'
    } }
  } as NavigationScreenProp<any, any>

  const defaults = {
    actions,
    navigation,
    state: store.getState()
  }

  it('displays a title based on the collection name', () => {
    expect(Screen.navigationOptions(defaults)).toEqual({
      title: 'Over the Hedge (Music from the Motion Picture)'
    })
  })

  it('renders a flat list of songs', () => {
    const tree = shallow(<Screen {...defaults} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders each song as a row', () => {
    const song = { trackId: 2 } as Song
    const tree = shallow(<Screen {...defaults} />)
    const render = tree.find('FlatList').prop('renderItem') as Function // tslint:disable-line
    const row = new ShallowWrapper(render({ item: song }), tree)
    expect(row.prop('song')).toBe(song)
  })

  it('uses trackId as key for each row', () => {
    const song = { trackId: 2 } as Song
    const tree = shallow(<Screen {...defaults} />)
    const keyExtractor = tree.find('FlatList').prop('keyExtractor') as Function // tslint:disable-line
    expect(keyExtractor(song)).toBe('2')
  })
})
