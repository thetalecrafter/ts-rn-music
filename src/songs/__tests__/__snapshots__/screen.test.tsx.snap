// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Songs screen renders a flat list of songs 1`] = `
<FlatList
  data={Array []}
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  onRefresh={[Function]}
  refreshing={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
